OpenSSL/crypto.py

```diff
@@ -25,13 +25,43 @@
 TYPE_DSA = _lib.EVP_PKEY_DSA
 
 
+
 class Error(Exception):
     """
     An error occurred in an `OpenSSL.crypto` API.
     """
 
 
+
+def _exception_from_context_error(exception_type, store_ctx):
+    """
+    Convert a :py:func:`OpenSSL.crypto.verify_cert` failure into a Python
+    exception.
+
+    When a call to native OpenSSL X509_verify_cert fails, additonal information
+    about the failure can be contained from the store context.
+    """
+
+    errors = [
+        _lib.X509_STORE_CTX_get_error(store_ctx._store_ctx),
+        _lib.X509_STORE_CTX_get_error_depth(store_ctx._store_ctx),
+        _native(_ffi.string(_lib.X509_verify_cert_error_string(_lib.X509_STORE_CTX_get_error(store_ctx._store_ctx)))),
+    ]
+    _x509 = _lib.X509_STORE_CTX_get_current_cert(store_ctx._store_ctx)
+    if _x509 != _ffi.NULL:
+      _cert = _lib.X509_dup(_x509)
+      pycert = X509.__new__(X509)
+      pycert._x509 = _ffi.gc(_cert, _lib.X509_free)
+    e = exception_type(errors)
+    e.certificate = pycert
```

Comment:
I think this is relevant. =P

> There are only two hard things in Computer Science: cache invalidation and naming things.
> 
> -- Phil Karlton

My gut wants to make us stick with the OpenSSL API and just name this thing a `X509StoreContextError`, but from poking around the OpenSSL code base I can only see this object being used for verification purposes (albeit, verification in lots of different places: `pkcs7`, `ocsp`, `x509_vfy`, ...). I think I should do more research about the naming and use of this thing. I'll let you know what I find.

As for the coding suggestions, they all make sense.